airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py

```diff
@@ -325,7 +325,7 @@ def create_legacy_to_per_partition_state_migration(
             raise ValueError(
                 f"LegacyToPerPartitionStateMigrations can only be applied on a DeclarativeStream with a SimpleRetriever. Got {type(declarative_stream.retriever)}"
             )
-        if not isinstance(declarative_stream.retriever.partition_router, SubstreamPartitionRouterModel):
+        if not isinstance(declarative_stream.retriever.partition_router, (SubstreamPartitionRouterModel, CustomPartitionRouterModel)):
```

Comment:
this won't work for any arbitrary CustomPartitionRouter, right?

there's a few more requirements
- It needs to be a substream partition router
- its parent_stream_config must be a dict

Is that right? can we encode this somehow?

Reply:
I believe `parent_stream_config` always dict if we use `CustomPartitionRouter` as it parsed in `create_custom_component`

Reply:
you're correct on the typing, but not all CustomPartitionRouter necessarily have a parent_stream_config.

We should at least verify its existence.

Fine to keep out of scope here, but this makes me think we're missing an abstraction. We should introduce CustomSubstreamPartitionRouters so we can enforce the presence of the parent_stream_config.